stdlib/public/core/Dictionary.swift

```diff
@@ -1520,6 +1520,61 @@ extension Dictionary: Hashable where Value: Hashable {
   }
 }
 
+extension Dictionary: _HasCustomAnyHashableRepresentation
+where Value: Hashable {
+  public func _toCustomAnyHashable() -> AnyHashable? {
+    return AnyHashable(_box: _DictionaryAnyHashableBox(self))
+  }
+}
+
+internal struct _DictionaryAnyHashableBox<Key: Hashable, Value: Hashable>
+  : _AnyHashableBox {
+  internal let _value: Dictionary<Key, Value>
+  internal let _canonical: Dictionary<AnyHashable, AnyHashable>
+
+  internal init(_ value: Dictionary<Key, Value>) {
+    self._value = value
+    self._canonical = value as Dictionary<AnyHashable, AnyHashable>
+  }
+
+  internal var _base: Any {
+    return _value
+  }
+
+  internal func _isEqual(to other: _AnyHashableBox) -> Bool? {
+    guard let other = other._asDictionary() else { return nil }
+    return _canonical == other
+  }
+
+  internal var _hashValue: Int {
+    return _canonical.hashValue
+  }
+
+  internal func _hash(into hasher: inout Hasher) {
+    _canonical.hash(into: &hasher)
+  }
+
+  internal func _rawHashValue(_seed: (UInt64, UInt64)) -> Int {
+    return _canonical._rawHashValue(seed: _seed)
+  }
+
+  internal func _unbox<T: Hashable>() -> T? {
+    return _value as? T
+  }
+
+  internal func _downCastConditional<T>(
+    into result: UnsafeMutablePointer<T>
+  ) -> Bool {
+    guard let value = _value as? T else { return false }
```

Comment:
Hmmm. If there were a fast path for `type(of: T) == Dictionary<AnyHashable, AnyHashable>.self`, would it obviate the need for `_asDictionary`, because `_isEqual` would be able to efficiently use  `_downCastConditional`? (Sorry, same question as before).

Reply:
Not sure I get this, but FWIW, it sure would be helpful if we could analyze the structure of T at runtime and detect that it's a Dictionary, extract its Key & Value types and perform an elementwise downcast/conversion from `_value`:

```swift
if case Dictionary<let K, let V> = T { ... }
```

There is a similar issue with `_SwiftNewtypeWrapper`, which uses it's `rawValue`'s `AnyHashable` representation, but can't easily be downcast from it. (`"hello" as AnyHashable as? NSAttributedString.Key` should be non-nil, but (as far as I know), `_ConcreteAnyHashableBox<String>._downCastConditional<T>` can't detect that the target type conforms to _SwiftNewtypeWrapper with `RawValue == String`, so we have to go through Foundation.